[Hackathon No.5] tril_indices OP

xiaoguoguo626807 opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
New features 

### PR changes
OPs 

### Describe
add tril_indices CPU kernel python API and unit test
fix issue：https://github.com/PaddlePaddle/Paddle/issues/40330
RFC https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20220323_api_design_for_tril_indices.md
中文文档PR https://github.com/PaddlePaddle/docs/pull/4599

xiaoguoguo626807 committed:
add tril_indices cpu kernal

xiaoguoguo626807 committed:
modify tril_indice cpu op

xiaoguoguo626807 committed:
Merge pull request #1 from xiaoguoguo626807/xgg-branch

merge tril_indices cpu op

xiaoguoguo626807 committed:
modify bug

xiaoguoguo626807 committed:
modify bug

xiaoguoguo626807 committed:
add tril_indices python api

xiaoguoguo626807 committed:
add tril_indices python api

xiaoguoguo626807 committed:
Merge pull request #2 from xiaoguoguo626807/xgg-branch

Xgg branch

xiaoguoguo626807 committed:
resolve conflict

xiaoguoguo626807 committed:
Merge pull request #3 from xiaoguoguo626807/xgg-branch

resolve conflict

xiaoguoguo626807 committed:
add tril_indices test

xiaoguoguo626807 committed:
Merge pull request #4 from xiaoguoguo626807/xgg-branch

add tril_indices test

xiaoguoguo626807 committed:
modify details

xiaoguoguo626807 committed:
add tril_indices.cu

xiaoguoguo626807 committed:
Merge pull request #5 from xiaoguoguo626807/xgg-branch

Xgg branch

xiaoguoguo626807 committed:
pythonapi pass

xiaoguoguo626807 committed:
Merge branch 'PaddlePaddle:develop' into develop

xiaoguoguo626807 committed:
save tril_indices

xiaoguoguo626807 committed:
solve conflict

xiaoguoguo626807 committed:
CPU tril_indices pass

xiaoguoguo626807 committed:
Merge pull request #6 from xiaoguoguo626807/pythonapi_pass

cpu tril_indices version

xiaoguoguo626807 committed:
delete vlog

xiaoguoguo626807 committed:
modify test_tril_indices_op.py

xiaoguoguo626807 committed:
solve conflict

xiaoguoguo626807 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into develop

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle?pullRequest=41639) <br/>All committers have signed the CLA.

xiaoguoguo626807 committed:
delete tril_indices_kernel.cc.swp

xiaoguoguo626807 committed:
delete tril_indice.cu

xiaoguoguo626807 committed:
modify code style

Ligoml commented:
1. 需要在describe中增加rfc链接和中文文档PR链接；
2. 待CI大部分通过后，你的pr会接收飞桨团队研发人员的review~

xiaoguoguo626807 committed:
add newline in creation.py

xiaoguoguo626807 committed:
modify creation.py linux newline

xiaoguoguo626807 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into develop

zhwesky2010 commented on paddle/fluid/operators/tril_indices_op.cc:
```diff
@@ -0,0 +1,73 @@
+ public:
+  void Make() override {
+    AddOutput("out","Tensor, the output tensor, with the shape (2,x),x bounded by [0,rows*cols])");
+    AddAttr<int>("rows", "int number, the input of tril_indices op").SetDefault(0);
```
这里的参数描述还是具体描述一下，不要都写 op输入

zhwesky2010 commented on paddle/phi/infermeta/nullary.cc:
```diff
@@ -115,4 +115,33 @@ void TruncatedGaussianRandomInferMeta(const std::vector<int>& shape,
+    tril_size += diff_row * cols;
+  }
+  std :: vector<int64_t> tmp ;
+  tmp.push_back(2);tmp.push_back(tril_size);
```
本地安装一下 pre-commit install；这里代码风格检查应该过不了

zhwesky2010 commented on paddle/phi/infermeta/nullary.cc:
```diff
@@ -115,4 +115,33 @@ void TruncatedGaussianRandomInferMeta(const std::vector<int>& shape,
+  if (diff_row > 0) {
+    tril_size += diff_row * cols;
+  }
+  std :: vector<int64_t> tmp ;
```
可以直接初始化一个 std::vector

zhwesky2010 commented on paddle/phi/kernels/cpu/tril_indices_kernel.cc:
```diff
@@ -0,0 +1,56 @@
+    int64_t tril_size = out_dims[1];
+    int64_t i = 0;
+
+    T r = std::max<int64_t>(0, -offset), c = 0;
```
安装pre-commit install 检查下这些地方的代码风格

zhwesky2010 commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,120 @@
+
+    def init_config(self):
+        self.attrs = {'rows': 4, 'cols': 4, 'offset': 0}
+        self.target = np.tril_indices(self.attrs['rows'],self.attrs['offset'],self.attrs['cols'])
```
这里返回的不是一个numpy.ndarray吗

xiaoguoguo626807 replied:
好像是tuple,不转的话静态图会报错

zhwesky2010 commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,120 @@
+class TestTrilIndicesOpCase1(TestTrilIndicesOp):
+    def init_config(self):
+        self.attrs = {'rows': 0, 'cols': 0, 'offset': 0}
+        #self.target = np.tril_indices(self.attrs['rows'],self.attrs['offset'],self.attrs['cols'])
```
这里可以删掉

zhwesky2010 commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,120 @@
+        def test_dygraph(self):
+            paddle.disable_static()
+            out =  paddle.tril_indices(4,4,2)
+            #print(out)
```
调试代码都可以删掉

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -1348,3 +1348,88 @@ def complex(real, imag, name=None):
+
+    if not isinstance(dtype, core.VarDesc.VarType):
+        dtype = convert_np_dtype_to_dtype_(dtype)
+    # if dtype != 2 and dtype != 3:
```
调试代码都可以删掉

zhwesky2010 commented on python/paddle/tensor/creation.py:
```diff
@@ -1348,3 +1348,88 @@ def complex(real, imag, name=None):
+        dtype = convert_np_dtype_to_dtype_(dtype)
+    # if dtype != 2 and dtype != 3:
+    #     dtype = 3
+    if in_dygraph_mode():
```
动态图这里需要最新的自动代码生成 python/paddle/utils/code_gen/api.yaml，然后使用_C_ops.final_state_tril_indices来调用

xiaoguoguo626807 replied:
这个地方是写yaml配置，然后根据配置自动生成吗。yaml里infer_meta和kernel项中的param是必须的吗，我看有的算子有有的没有，好像没有和args对齐

zhwesky2010 replied:
会通过yaml配置自动生成

zhwesky2010 reviewed (commented):
后面会有GPU资源，kernel.cu完成后再加上相应单测

xiaoguoguo626807 committed:
delete annotation

xiaoguoguo626807 committed:
check code style

xiaoguoguo626807 committed:
check .py style add final_state??

xiaoguoguo626807 committed:
modify code style

xiaoguoguo626807 committed:
Merge branch 'PaddlePaddle:develop' into develop

zhwesky2010 commented on python/paddle/utils/code_gen/api.yaml:
```diff
@@ -1,3 +1,15 @@
+# - api : tril_indices
```
这里注释了，没有进行自动代码生成吗

xiaoguoguo626807 replied:
这个之前测试的时候有点问题，后面提交的版本动态图支持了新旧两个版本

zhwesky2010 replied:
这里没有yaml自动生成，_C_ops.final_state_tril_indices是怎么调用的呢？

xiaoguoguo626807 replied:
测试文件没有走那个函数，所以通过了，新的提交修改了测试函数，也添加了gpu测试

xiaoguoguo626807 committed:
Merge branch 'PaddlePaddle:develop' into develop

xiaoguoguo626807 committed:
add gpu_tril_indices

xiaoguoguo626807 committed:
Merge branch 'develop' of https://github.com/xiaoguoguo626807/Paddle into develop

xiaoguoguo626807 committed:
modify gpu_compiled_juage

xiaoguoguo626807 committed:
modify gpu judge

xiaoguoguo626807 committed:
code style

zhwesky2010 reviewed (commented):
代码看起来没什么问题了，能通过所有CI就可以了

xiaoguoguo626807 commented:
> 代码看起来没什么问题了，能通过所有CI就可以了

好的

zhwesky2010 reviewed (dismissed):
LGTM

Ligoml commented on python/paddle/tensor/creation.py:
```diff
@@ -1739,3 +1739,97 @@ def complex(real, imag, name=None):
+
+def tril_indices(rows, cols, offset=0, dtype='int64'):
+    """
+    This op returns the indices of the lower triangular part of the 2-D matrix 
```
参考中文文档review意见修改，需要保持中英文文档内容一致，参数数量一致~

Ligoml replied:
和中文文档一致，删掉「This op」

Ligoml commented on python/paddle/tensor/creation.py:
```diff
@@ -1739,3 +1739,97 @@ def complex(real, imag, name=None):
+    Examples:
+        .. code-block:: python
+
+            import numpy as np
```
不需要import numpy，这里需要删掉

shjNT commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,145 @@
+            out = paddle.tril_indices(4, None, 2)
+        expected_result = np.tril_indices(4, 2)
+        self.assertEqual((out.numpy() == expected_result).all(), True)
+
```
建议单测覆盖offset<0的情况

xiaoguoguo626807 replied:
已添加

xiaoguoguo626807 committed:
add test example

xiaoguoguo626807 committed:
modify english document

modify english document

modify english document

modify document

modify document

Ligoml commented on python/paddle/tensor/creation.py:
```diff
@@ -1739,3 +1739,95 @@ def complex(real, imag, name=None):
+        Tensor: Results of the indices of lower triangular part of a rows * cols matrix,
+        where the first row contains row coordinates of and the second row contains column coordinates.
+
+    Raises:
```
raises删掉，报错里有就可以了，文档不需要展示

Ligoml commented on python/paddle/tensor/creation.py:
```diff
@@ -1739,3 +1739,95 @@ def complex(real, imag, name=None):
+    Args:
+        rows (int): The input x which is a int number describe the number of row of the matrix.
+        cols (int): The input x which is a int number describe the number of col of the matrix.    
+        offset (int, optional): The offset to consider, default value is 0.
```
英文预览也不太对
<img width="892" alt="image" src="https://user-images.githubusercontent.com/39876205/165687318-9ae8976a-6737-4f15-91fc-2c7284e9cf9a.png">

xiaoguoguo626807 replied:
为什么：会变成- ，英文文档在哪里查看呢

Ligoml replied:
中英文文档都在docs的pr里预览，[参考文档](https://github.com/PaddlePaddle/docs/wiki/%5BBeta%5D%E9%A3%9E%E6%A1%A8%E6%96%87%E6%A1%A3%E9%A2%84%E8%A7%88%E5%B7%A5%E5%85%B7)

jeff41404 commented on paddle/phi/infermeta/nullary.cc:
```diff
@@ -115,4 +115,27 @@ void TruncatedGaussianRandomInferMeta(const std::vector<int>& shape,
+void TrilIndicesInferMeta(
+    int rows, int cols, int offset, DataType dtype, MetaTensor* out) {
+  // number of elements in the first row of the tril,bounded by [0, cols]
+  auto m_first_row =
```
use `n_first_row` is better?  the form of variable name is unified with `n_row_all` and `n_row_trapezoid` below

xiaoguoguo626807 replied:
done

jeff41404 commented on paddle/phi/infermeta/nullary.cc:
```diff
@@ -115,4 +115,27 @@ void TruncatedGaussianRandomInferMeta(const std::vector<int>& shape,
+  auto m_first_row =
+      offset > 0 ? std::min<int64_t>(cols, 1 + offset) : rows + offset > 0;
+  // number of elements in the last row of the tril, bounded by [0, cols]
+  auto m_last_row =
```
use `n_last_row` is better? the form of variable name is unified with `n_row_all` and `n_row_trapezoid` below

xiaoguoguo626807 committed:
modify pram name

xiaoguoguo626807 committed:
modify pram name

Ligoml reviewed (dismissed):
LGTM for docs

jeff41404 reviewed (dismissed):
LGTM

XiaoguangHu01 commented on python/paddle/tensor/creation.py:
```diff
@@ -1739,3 +1739,92 @@ def complex(real, imag, name=None):
     return out
+
+
+def tril_indices(rows, cols, offset=0, dtype='int64'):
```
这里参数名用单数稍好一些，rows -> row, cols -> col
复数一般用来表示列表，比如，slice这个api中，axes, starts, ends都是用来接受列表的输入
https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/slice_cn.html#slice

xiaoguoguo626807 replied:
done

xiaoguoguo626807 committed:
modify pram

xiaoguoguo626807 committed:
fix conflict

zhwesky2010 commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,153 @@
+        expected_result = np.tril_indices(4, 2)
+        self.assertEqual((out.numpy() == expected_result).all(), True)
+
+    def test_default_CPU(self):
```
测默认值就不分CPU和GPU吧，节省一点case时间，目前有单测超时的可能（15s）

xiaoguoguo626807 replied:
已改，但是对测试总时间影响不大

zhwesky2010 commented on python/paddle/fluid/tests/unittests/test_tril_indices_op.py:
```diff
@@ -0,0 +1,153 @@
+            with paddle.static.program_guard(paddle.static.Program(),
+                                             paddle.static.Program()):
+                data1 = paddle.tril_indices(4, 4, -1)
+                exe1 = paddle.static.Executor(place)
```
每个case测一个值就可以，或者静态图 为 正offset，动态图为 负offset，节省一些单测时间

xiaoguoguo626807 replied:
同上

zhwesky2010 replied:
看下超时能不能过吧，如果还超时，那就需要再继续减少case了，目前不允许单测超过15s

xiaoguoguo626807 committed:
reduce test ex

zhwesky2010 reviewed (approved):
LGTM

XiaoguangHu01 reviewed (approved):
LGTM
